modules/gdnative/nativearvr/arvr_interface_gdnative.h

```diff
@@ -0,0 +1,88 @@
+/*************************************************************************/
+/* Copyright (c) 2007-2017 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2017 Godot Engine contributors (cf. AUTHORS.md)    */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef ARVR_INTERFACE_GDNATIVE_H
+#define ARVR_INTERFACE_GDNATIVE_H
+
+#include "modules/gdnative/gdnative.h"
+#include "servers/arvr/arvr_interface.h"
+
+/**
+	@authors Hinsbart & Karroffel & Mux213
+
+	This subclass of our AR/VR interface forms a bridge to GDNative.
+*/
+
+class ARVRInterfaceGDNative : public ARVRInterface {
+	GDCLASS(ARVRInterfaceGDNative, ARVRInterface);
+
+	void cleanup();
+
+protected:
+	Ref<GDNative> library;
+	void *data;
+
+	static void _bind_methods();
+
+public:
+	/** general interface information **/
+	ARVRInterfaceGDNative();
+	~ARVRInterfaceGDNative();
+
+	void set_gdnative_library(Ref<GDNativeLibrary> p_library);
+
+	virtual StringName get_name() const;
+	virtual int get_capabilities() const;
+
+	virtual bool is_initialized();
+	virtual bool initialize();
+	virtual void uninitialize();
+
+	/** specific to AR **/
+	virtual bool get_anchor_detection_is_enabled() const;
+	virtual void set_anchor_detection_is_enabled(bool p_enable);
+
+	/** rendering and internal **/
+	virtual Size2 get_recommended_render_targetsize();
+	virtual bool is_stereo();
+	virtual Transform get_transform_for_eye(ARVRInterface::Eyes p_eye, const Transform &p_cam_transform);
+
```

Comment:
All the above functions call into a library, that's the way you can swap out behaviors. I don't think those methods should be virtual, it's only bad for performance considering that this class will probably never be subtyped anyway (because who needs to subtype something that literally just delegates all the work).

Same for the functions below.

If there are plans where it would be wise to inherit from this class then of course that's okay. If not then those `virtual`s are only bad and not good 😛

Reply:
They already are virtual. Remember ARVRInterfaceGDNative implements the behaviour set by ARVRInterface which is our template class. They are also core interfaces such as ARKit and mobile VR. Especially ARKIt requires some interesting changes to work that I'm not sure can be done as a native module.

Reply:
Ah yes, I forgot that.. :D